### ISSUE TITLE
Dockerfile for faster installation

### ISSUE THREAD

**Issue by dmitry** [drolbr/Overpass-API#318]
Any specific reason why there are no Dockerfile, that allows to setup and run overpass API with just few commands?

It took me some time to setup Docker image and run container: https://github.com/dmitry/docker-overpass-api

Even after that I don't feel it's the final solution, that allows to query everything. Would be great if the core source of overpass API will support Dockerfile that allows to setup docker container in a few minutes, because if something went wrong it could take half a day or even days to rebuild everything from scratch.

**Comment by mmd-osm** [comment]
Why don't you sync up with https://github.com/mediasuitenz/docker-overpass-api where your repo obviously originates from and create a common Docker image?

**Comment by dmitry** [comment]
@mmd-osm took most of the data from https://github.com/vespakoen/ontheweb/tree/master/overpass-api, but still had an issue with areas. Why not to push Dockerfile directly into the  https://github.com/drolbr/Overpass-API ?

**Comment by mmd-osm** [comment]
@vespakoen, @mediasuitenz: are you maybe interested in creating a best practise Dockerfile?

**Comment by dmitry** [comment]
`areas.osm3s` - that's a good suggestion, I don't like this idea too. It was fast solution, as in the archive there are no rules path with areas data.

Is there are any issues with overpass execution while optimization during compilation was used?

**Comment by mmd-osm** [comment]
Yes. we had segmentation faults with -O3 for some versions of gcc during test runs. I think -O2 is still safe to use (didn't have issues with it), Roland removed the optimization altogether in the  docs.

**Comment by dmitry** [comment]
@mmd-osm regarding the `areas-osm3s` haven't noticed it now in the 0.7.53, `rules` were missing in the previous release (http://dev.overpass-api.de/releases/osm-3s_v0.7.52.tar.gz). Thanks for the `-O2` suggestion!

**Comment by mmd-osm** [comment]
It would be cool to have some script where you can decide the compression (zlib, lz4, no compression) / map compression / meta settings / area to be created, etc. There are plenty of different options, which would be worthwhile configuring.

Also the question, how updates should be handled (automatic for planet, no updates, ...).

**Comment by dmitry** [comment]
@mmd-osm I just started to work with overpass API, haven't dig too deeply into the settings. But all the current docker images I found on the github somehow broken.

Regarding automatic updates, looks like it's very CPU and memory intensitive task. Should be easily switchable, may be document something about switching it on / off using `docker exec` in the current container instance, so it could be turned on while overpass is working, without rebuilding everything from scratch.

**Comment by dmitry** [comment]
Regarding `O3` - there are many resources that suggests to use `O3`:

- http://overpass-api.de/full_installation.html
- http://wiki.openstreetmap.org/wiki/Overpass_API/Installation
- http://overpass-api.de/no_frills.html

**Comment by mmd-osm** [comment]
Right, that's due to the main instance still running on 0.7.52. This has been changed in this [commit](https://github.com/drolbr/Overpass-API/commit/50d2e9c5b488f802c8fdbdeaaf381567e4059447), which is part of 0.7.53.
Removing optimization altogether is a bit pessimistic, imho. There's also another issue for it: #316

**Comment by Frankkkkk** [comment]
Mentionned my dockerfile on the mailing list. It is [here](https://github.com/Frankkkkk/docker-overpass-api). It does nearly everything except XAPI and incremental areas. What do you think ?

**Comment by mmd-osm** [comment]
@Frankkkkk : some of my previous comments also apply to your Docker file: 
* dedicated user instead of root
* use of `-O3` compilation option is currently discouraged, 
* use only released version (master branch should not be used).
* Don't download whole planet and store in docker image, better use PBF + osmconvert and store outside of docker
* Consider option for cloning, rather than setting up db from scratch (only take 3-4 hours on fast network)

**Comment by Frankkkkk** [comment]
Hi,

- Root doesn't matter: it is inside a docker container so even if there is a security breach, the master host is not vulnerable
- will fix `-O3`
- will see for last version
- the whole planet is not stored in the docker image. It is downloaded in the host and deleted after the import
- for the clone, I should check that someday. I wanted first the option to be independent (i.e. start from scratch). It implies to use the same version as the cloning server. I will see about that.

Cheers

**Comment by mmd-osm** [comment]
Regarding security: 

> Docker containers are, by default, quite secure; especially if you take care of running your processes inside the containers as non-privileged users (i.e., non-root).

Source: https://docs.docker.com/engine/security/security/

We advocate to use a non-root account everywhere in the installation docs. I'd expect exactly the same least privilege principle for a Docker image as well.

**Comment by Frankkkkk** [comment]
But also: "This means that even if an intruder manages to escalate to root within a container, it will be much harder to do serious damage, or to escalate to the host.".

Anyways, I don't want to spend more energy on this. I agree that its prettier to have its own user. If I've got the time, I'll do it.

**Comment by Frankkkkk** [comment]
Okay, I've updated some of your points (user, O3, last release).  It is [here](https://github.com/Frankkkkk/docker-overpass-api).

What do you think ?

**Comment by Frankkkkk** [comment]
Some news ? ;-)

**Comment by gy-mate** [comment]
A maintained Docker image is available in this repo (and on Docker Hub): https://github.com/wiktorn/Overpass-API

This issue can probably be closed.

**Comment by b1tw153** [comment]
There's also https://github.com/b1tw153/Overpass-API or the [b1w153/overpass-api](https://hub.docker.com/r/b1tw153/overpass-api) image on Docker Hub for an alternate container build.
